src/net/atlas/db.rs

```diff
@@ -0,0 +1,379 @@
+            return Err(db_error::ReadOnly);
+        }
+
+        let tx = tx_begin_immediate(&mut self.conn)?;
+        Ok(tx)
+    }
+
+    pub fn get_minmax_heights_window_for_page_index(
+        &self,
+        oldest_page_index: u32,
+        newest_page_index: u32,
+    ) -> Result<(u64, u64), db_error> {
+        let qry = "SELECT MIN(block_height) as min, MAX(block_height) as max FROM attachment_instances WHERE page_index >= ?1 AND page_index <= ?2".to_string();
+        let args = [
+            &oldest_page_index as &dyn ToSql,
+            &newest_page_index as &dyn ToSql,
+        ];
+        let mut stmt = self.conn.prepare(&qry)?;
+        let mut rows = stmt.query(&args)?;
+        match rows.next() {
+            Some(Ok(row)) => {
+                let min: i64 = row.get_checked("min")?;
+                let max: i64 = row.get_checked("max")?;
+                Ok((min as u64, max as u64))
+            }
+            _ => Err(db_error::NotFoundError),
+        }
+    }
+
+    pub fn get_attachments_available_at_pages_indexes(
+        &self,
+        pages_indexes: &Vec<u32>,
+        blocks_ids: &Vec<StacksBlockId>,
+    ) -> Result<Vec<Vec<u8>>, db_error> {
+        let mut pages = vec![];
+        for page_index in pages_indexes {
+            let page = self.get_attachments_missing_at_page_index(*page_index, blocks_ids)?;
+            let mut bit_vector = vec![];
+            for (_index, is_attachment_missing) in page.iter().enumerate() {
+                // todo(ludo): use a bitvector instead
+                bit_vector.push(if *is_attachment_missing { 0 } else { 1 });
+            }
+            pages.push(bit_vector);
+        }
+        Ok(pages)
+    }
+
+    pub fn get_attachments_missing_at_page_index(
+        &self,
+        page_index: u32,
+        blocks_ids: &Vec<StacksBlockId>,
+    ) -> Result<Vec<bool>, db_error> {
+        // todo(ludo): unable to build a compiled stmt with rusqlite that includes a WHERE ... IN () clause - investigate carray.
+        let ancestor_tree_sql = blocks_ids
+            .iter()
+            .map(|block_id| format!("'{}'", block_id))
+            .collect::<Vec<String>>()
+            .join(", ");
+        let qry = format!("SELECT is_available FROM attachment_instances WHERE page_index = {} AND block_id IN ({}) ORDER BY position_in_page ASC", page_index, ancestor_tree_sql);
+        let rows = query_rows::<i64, _>(&self.conn, &qry, NO_PARAMS)?;
+        let res = rows.iter().map(|r| *r == 0).collect::<Vec<bool>>();
+        println!("get_attachments_missing_at_page_index res: {:?}", res);
+        Ok(res)
+    }
+
+    pub fn insert_new_inboxed_attachment(
+        &mut self,
+        attachment: Attachment,
+    ) -> Result<(), db_error> {
+        // Do we already have an entry (proceessed or unprocessed) for this attachment? - todo(ludo) think more about this
+        let qry = "SELECT count(*) FROM inboxed_attachments WHERE hash = ?1".to_string();
```

Comment:
The `COUNT(*)` function in sqlite has _Omega(n)_ time complexity for _n_ rows.  You should try to avoid using it if at all possible, since this function will take longer and longer to run as there are more and more attachments.

Reply:
It's probably fine if you store the same attachment twice (it's asymptotically preferable to do this), as long as two things are true:

* The attachment is content-addressed -- i.e. the hash160 of the `Attachment`'s content is _always_ equal to its hash (related, you may want to just drop the `.hash` field, and replace it with a method `.hash()` that just calculates the hash160 -- then, it'll be _impossible_ to create an `Attachment` whose hash and content differ).
* The attachment's `.content` field has a fixed upper-bound on its size.

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/2013/commits/54e632f3f10aad4e631f5a2fe35575228e452f46